File: acp.go
package main

import (
	"bufio"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"os/exec"
	"sync"
)

// Update is one session/update notification, flattened to what the UI needs.
type Update struct {
	Kind     string // sessionUpdate, e.g. agent_message_chunk, tool_call, plan
	Text     string // for message and thought chunks
	Title    string // for tool calls
	ToolPath string // first location of a tool call; blast radius uses these
	Raw      json.RawMessage
}

type PermOption struct {
	ID   string `json:"optionId"`
	Name string `json:"name"`
	Kind string `json:"kind"`
}

// PermissionReq is a blocked agent. Sending an option ID on Reply unblocks
// it; closing Reply without a send cancels.
type PermissionReq struct {
	Title   string
	Options []PermOption
	Reply   chan string
}

type Client struct {
	Updates chan Update
	Perms   chan PermissionReq

	cmd  *exec.Cmd
	out  *json.Encoder
	in   *bufio.Scanner
	mu   sync.Mutex
	next int
	// pending maps an outgoing request id to the channel its response lands on.
	pending map[int]chan rpcResponse
	writeMu sync.Mutex
	closed  chan struct{}
}

type rpcResponse struct {
	Result json.RawMessage
	Err    error
}

type rpcFrame struct {
	JSONRPC string          `json:"jsonrpc"`
	ID      json.RawMessage `json:"id,omitempty"`
	Method  string          `json:"method,omitempty"`
	Params  json.RawMessage `json:"params,omitempty"`
	Result  json.RawMessage `json:"result,omitempty"`
	Error   *rpcError       `json:"error,omitempty"`
}

type rpcError struct {
	Code    int    `json:"code"`
	Message string `json:"message"`
}

// Dial starts the agent process and completes the ACP handshake.
func Dial(ctx context.Context, argv []string, cwd string) (*Client, error) {
	cmd := exec.CommandContext(ctx, argv[0], argv[1:]...)
	cmd.Dir = cwd
	stdin, err := cmd.StdinPipe()
	if err != nil {
		return nil, err
	}
	stdout, err := cmd.StdoutPipe()
	if err != nil {
		return nil, err
	}
	if err := cmd.Start(); err != nil {
		return nil, err
	}
	c := &Client{
		Updates: make(chan Update, 128),
		Perms:   make(chan PermissionReq, 8),
		cmd:     cmd,
		out:     json.NewEncoder(stdin),
		in:      bufio.NewScanner(stdout),
		pending: map[int]chan rpcResponse{},
		closed:  make(chan struct{}),
	}
	// Agent output is line-delimited JSON, and a long tool result is a long
	// line. The default 64K scanner limit would truncate it into a parse error.
	c.in.Buffer(make([]byte, 0, 64*1024), 8*1024*1024)
	go c.readLoop()

	if _, err := c.call("initialize", map[string]any{
		"protocolVersion": 1,
		"clientCapabilities": map[string]any{
			"fs": map[string]any{"readTextFile": false, "writeTextFile": false},
		},
	}); err != nil {
		cmd.Process.Kill()
		return nil, fmt.Errorf("initialize: %w", err)
	}
	return c, nil
}

func (c *Client) NewSession(cwd string) (string, error) {
	res, err := c.call("session/new", map[string]any{
		"cwd":        cwd,
		"mcpServers": []any{},
	})
	if err != nil {
		return "", err
	}
	var out struct {
		SessionID string `json:"sessionId"`
	}
	if err := json.Unmarshal(res, &out); err != nil {
		return "", err
	}
	return out.SessionID, nil
}

// Prompt sends a turn and blocks until the agent finishes it. Updates and
// permission requests arrive on the channels while it is in flight, so it
// must be called from its own goroutine.
func (c *Client) Prompt(sessionID, text string) (string, error) {
	res, err := c.call("session/prompt", map[string]any{
		"sessionId": sessionID,
		"prompt":    []any{map[string]any{"type": "text", "text": text}},
	})
	if err != nil {
		return "", err
	}
	var out struct {
		StopReason string `json:"stopReason"`
	}
	json.Unmarshal(res, &out)
	return out.StopReason, nil
}

func (c *Client) Cancel(sessionID string) error {
	return c.notify("session/cancel", map[string]any{"sessionId": sessionID})
}

func (c *Client) Close() error {
	select {
	case <-c.closed:
	default:
		close(c.closed)
	}
	if c.cmd.Process != nil {
		c.cmd.Process.Kill()
	}
	return c.cmd.Wait()
}

func (c *Client) call(method string, params any) (json.RawMessage, error) {
	c.mu.Lock()
	c.next++
	id := c.next
	ch := make(chan rpcResponse, 1)
	c.pending[id] = ch
	c.mu.Unlock()

	if err := c.write(rpcFrame{
		JSONRPC: "2.0",
		ID:      json.RawMessage(fmt.Sprint(id)),
		Method:  method,
		Params:  mustJSON(params),
	}); err != nil {
		return nil, err
	}
	select {
	case r := <-ch:
		return r.Result, r.Err
	case <-c.closed:
		return nil, io.ErrUnexpectedEOF
	}
}

func (c *Client) notify(method string, params any) error {
	return c.write(rpcFrame{JSONRPC: "2.0", Method: method, Params: mustJSON(params)})
}

func (c *Client) write(f rpcFrame) error {
	c.writeMu.Lock()
	defer c.writeMu.Unlock()
	return c.out.Encode(f)
}

// readLoop is the only reader of the agent's stdout. It routes three shapes:
// a response (id, no method), an incoming request (id and method), and a
// notification (method, no id).
func (c *Client) readLoop() {
	defer func() {
		select {
		case <-c.closed:
		default:
			close(c.closed)
		}
		close(c.Updates)
		close(c.Perms)
	}()

	for c.in.Scan() {
		var f rpcFrame
		if err := json.Unmarshal(c.in.Bytes(), &f); err != nil {
			continue // a line we cannot parse is not worth dying over
		}
		switch {
		case f.Method == "" && len(f.ID) > 0:
			c.deliver(f)
		case f.Method == "session/request_permission":
			c.handlePermission(f)
		case f.Method == "session/update":
			c.handleUpdate(f)
		}
	}
}

func (c *Client) deliver(f rpcFrame) {
	var id int
	if err := json.Unmarshal(f.ID, &id); err != nil {
		return
	}
	c.mu.Lock()
	ch := c.pending[id]
	delete(c.pending, id)
	c.mu.Unlock()
	if ch == nil {
		return
	}
	if f.Error != nil {
		ch <- rpcResponse{Err: fmt.Errorf("agent error %d: %s", f.Error.Code, f.Error.Message)}
		return
	}
	ch <- rpcResponse{Result: f.Result}
}

func (c *Client) handleUpdate(f rpcFrame) {
	var p struct {
		Update struct {
			SessionUpdate string `json:"sessionUpdate"`
			Content       struct {
				Text string `json:"text"`
			} `json:"content"`
			Title     string `json:"title"`
			Locations []struct {
				Path string `json:"path"`
			} `json:"locations"`
		} `json:"update"`
	}
	if err := json.Unmarshal(f.Params, &p); err != nil {
		return
	}
	u := Update{
		Kind:  p.Update.SessionUpdate,
		Text:  p.Update.Content.Text,
		Title: p.Update.Title,
		Raw:   f.Params,
	}
	if len(p.Update.Locations) > 0 {
		u.ToolPath = p.Update.Locations[0].Path
	}
	// Unknown kinds ride through as-is. The UI ignores what it does not
	// recognize; dying here would break on every agent release.
	select {
	case c.Updates <- u:
	case <-c.closed:
	}
}

func (c *Client) handlePermission(f rpcFrame) {
	// The toolCall here is an ACP ToolCallUpdate, in which only toolCallId is
	// required — title, kind and locations are all optional.
	var p struct {
		ToolCall struct {
			Title      string `json:"title"`
			Kind       string `json:"kind"`
			ToolCallID string `json:"toolCallId"`
		} `json:"toolCall"`
		Options []PermOption `json:"options"`
	}
	json.Unmarshal(f.Params, &p)

	req := PermissionReq{
		Title:   permTitle(p.ToolCall.Title, p.ToolCall.Kind, p.ToolCall.ToolCallID),
		Options: p.Options,
		Reply:   make(chan string, 1),
	}
	select {
	case c.Perms <- req:
	case <-c.closed:
		return
	}

	// Answer on a goroutine: the read loop must not block, or the agent's
	// own output would stall behind the human.
	go func() {
		outcome := map[string]any{"outcome": "cancelled"}
		select {
		case id, ok := <-req.Reply:
			if ok {
				outcome = map[string]any{"outcome": "selected", "optionId": id}
			}
		case <-c.closed:
			return
		}
		c.write(rpcFrame{
			JSONRPC: "2.0",
			ID:      f.ID,
			Result:  mustJSON(map[string]any{"outcome": outcome}),
		})
	}()
}

// permTitle never returns an empty string. A permission prompt is the one
// place the user authorizes an action they cannot see, so a blank label would
// be asking them to approve nothing in particular. Degrade to the tool kind,
// then to the call id, then to an explicit warning — but always say something.
func permTitle(title, kind, id string) string {
	if title != "" {
		return title
	}
	if kind != "" && id != "" {
		return kind + " (" + id + ")"
	}
	if kind != "" {
		return kind
	}
	if id != "" {
		return "tool call " + id
	}
	return "UNIDENTIFIED tool call — the agent sent no description"
}

func mustJSON(v any) json.RawMessage {
	b, err := json.Marshal(v)
	if err != nil {
		panic(err) // only reachable with a non-serializable literal in our own code
	}
	return b
}

File: src/acp.ts
import { spawn } from "node:child_process";
import { relative, sep } from "node:path";
import { Readable, Writable } from "node:stream";
import { setTimeout as delay } from "node:timers/promises";
import {
  ClientSideConnection,
  ndJsonStream,
  PROTOCOL_VERSION,
  type Client,
  type ContentBlock,
  type PermissionOption,
  type RequestPermissionRequest,
  type RequestPermissionResponse,
  type SessionModeState,
  type SessionNotification,
  type SessionUpdate,
  type ToolCallUpdate,
} from "@agentclientprotocol/sdk";

// What the UI needs from a session/update notification, plus the raw payload.
export interface AgentUpdate {
  kind: SessionUpdate["sessionUpdate"];
  text?: string;
  // Set on tool_call/tool_call_update. toolCallId identifies the tool call
  // an update belongs to, so a client can merge updates into one entry
  // instead of rendering each as a separate line. toolTitle/toolKind are
  // only present when this particular update carries them (ACP updates are
  // patches: a field's absence means "unchanged", not "cleared").
  toolCallId?: string;
  toolTitle?: string;
  toolKind?: string;
  toolPath?: string;
  // "pending" | "in_progress" | "completed" | "failed". A tool that failed
  // renders differently from one that worked, so this is not cosmetic: an
  // agent thrashing through failing commands looks identical to one making
  // progress without it.
  toolStatus?: string;
  // Set when kind is "current_mode_update": the agent switched modes on its
  // own. The UI must reflect this rather than keep showing a stale mode.
  modeId?: string;
  // Set when kind is "usage_update": how full the agent's context window is
  // and what the session has cost so far. ACP marks this update UNSTABLE and
  // adapters may never send it, so everything downstream must treat its
  // absence as "unknown" and show nothing — never a zero.
  usage?: { used: number; size: number; cost?: { amount: number; currency: string } };
  raw: SessionUpdate;
}

// What the UI needs to render a permission prompt. `title` is sanitized and
// never blank. Match `options` on `kind` (PermissionOptionKind) — ids and
// array order are adapter-specific and unstable.
export interface PermissionRequest {
  title: string;
  options: PermissionOption[];
  toolCall: ToolCallUpdate;
}

export interface AgentSession {
  readonly sessionId: string;
  // The session's mode state as of the last newSession/setMode/
  // current_mode_update, or null if this adapter doesn't report modes at
  // all. The caller (not this module) decides whether to pin a default.
  readonly modes: SessionModeState | null;
  prompt(text: string): Promise<string>;
  cancel(): Promise<void>;
  // Requests a mode change. Throws if the adapter rejects it (e.g. unknown
  // modeId). On success, `modes.currentModeId` reflects the new mode.
  setMode(modeId: string): Promise<void>;
  close(): Promise<void>;
}

export interface SessionOptions {
  cwd: string;
  onUpdate(u: AgentUpdate): void;
  onPermission(req: PermissionRequest): Promise<string | null>;
  argv?: string[];
  // How long to wait for the adapter to answer initialize/newSession before
  // giving up. A dead or never-spawned adapter is caught by childClosed
  // (close/error events) long before this fires; this only covers one that
  // neither replies nor exits. Overridable so tests can exercise it without
  // a real multi-second wait.
  startupTimeoutMs?: number;
}

const DEFAULT_ARGV = ["npx", "-y", "@zed-industries/claude-code-acp"];
const DEFAULT_STARTUP_TIMEOUT_MS = 20_000;

// startupTimeout rejects after `ms` unless something else settles first.
// unref()'d so a pending one never keeps the process alive, and — since
// Promise.race attaches its own handler to every promise it's given,
// including the one that loses — this never produces an unhandled
// rejection either, so there's no need to cancel it explicitly.
function startupTimeout(ms: number): Promise<never> {
  return new Promise((_resolve, reject) => {
    const timer = setTimeout(() => {
      reject(new Error(`claude-code-acp did not respond within ${ms}ms of starting`));
    }, ms);
    timer.unref();
  });
}

// C0/C1 control characters. The agent controls every string here, and an
// ANSI escape in a permission dialog is an obvious attack.
const CONTROL_CHARS = /[\x00-\x1f\x7f-\x9f]/g;

// Strip control chars; a whitespace-only result counts as absent.
function clean(s: string | null | undefined): string | null {
  if (s == null) return null;
  const stripped = s.replace(CONTROL_CHARS, "").trim();
  return stripped === "" ? null : stripped;
}

// Never blank: title, then kind, then the tool call id, then an explicit
// marker so a bad payload is visible instead of read as an empty prompt.
function permissionTitle(toolCall: ToolCallUpdate): string {
  return (
    clean(toolCall.title) ??
    clean(toolCall.kind ?? null) ??
    clean(toolCall.toolCallId) ??
    "UNIDENTIFIED tool call"
  );
}

// Blast radius compares against repo-relative `git grep -l` output, so
// relativize against cwd. A location outside cwd stays absolute rather than
// emit a leading "../" that could never match; ".." is checked as a whole
// path segment since "..foo" is a real filename, not an escape.
function relativizePath(cwd: string, absPath: string): string {
  const rel = relative(cwd, absPath);
  return rel === ".." || rel.startsWith(".." + sep) ? absPath : rel;
}

function textOf(content: ContentBlock): string | undefined {
  return content.type === "text" ? content.text : undefined;
}

function toUpdate(cwd: string, n: SessionNotification): AgentUpdate {
  const u = n.update;
  const out: AgentUpdate = { kind: u.sessionUpdate, raw: u };
  switch (u.sessionUpdate) {
    case "user_message_chunk":
    case "agent_message_chunk":
    case "agent_thought_chunk":
      out.text = textOf(u.content);
      break;
    case "tool_call":
    case "tool_call_update":
      out.toolCallId = u.toolCallId;
      if (u.title != null) out.toolTitle = u.title;
      if (u.kind != null) out.toolKind = u.kind;
      if (u.status != null) out.toolStatus = u.status;
      if (u.locations && u.locations.length > 0) {
        out.toolPath = relativizePath(cwd, u.locations[0].path);
      }
      break;
    case "current_mode_update":
      out.modeId = u.currentModeId;
      break;
    case "usage_update":
      if (typeof u.used === "number" && typeof u.size === "number" && u.size > 0) {
        out.usage = { used: u.used, size: u.size };
        if (u.cost != null) out.usage.cost = { amount: u.cost.amount, currency: u.cost.currency };
      }
      break;
  }
  return out;
}

// Launches claude-code-acp and speaks ACP over its stdio via the SDK. We
// implement only the two Client methods the SDK requires; everything else
// (the read loop, the write queue, request/response correlation) is its job.
export async function startSession(opts: SessionOptions): Promise<AgentSession> {
  const argv = opts.argv ?? DEFAULT_ARGV;

  // claude-code-acp refuses to start under these: "Claude Code cannot be
  // launched inside another Claude Code session."
  const env = { ...process.env };
  delete env.CLAUDECODE;
  delete env.CLAUDE_CODE_SSE_PORT;

  const child = spawn(argv[0], argv.slice(1), {
    cwd: opts.cwd,
    env,
    // Make child its own process group leader so close() can signal the
    // whole tree, not just this immediate process. argv is commonly
    // ["npx", "-y", pkg]; npx execs the real agent through a shell wrapper
    // that does not forward signals to it, so killing only `child` leaves
    // the actual agent process orphaned and running. POSIX-only — win32 has
    // no process-group signal (negative pid), so it falls back to killing
    // just this process there (see killTree).
    detached: process.platform !== "win32",
  });

  // Drain stderr unconditionally so a chatty child never blocks on a full
  // pipe; keep a short tail to enrich the error if it exits before replying.
  let stderrTail = "";
  child.stderr.on("data", (chunk: Buffer) => {
    stderrTail = (stderrTail + chunk.toString("utf8")).slice(-4096);
  });

  // The SDK never rejects a pending request when the stream ends — it just
  // hangs (see Connection#sendRequest/#receive in dist/acp.js). If the child
  // dies mid-turn, this is what makes `prompt()` reject instead of hanging —
  // and, raced against initialize()/newSession() below, what makes startup
  // fail loudly instead of hanging forever when the adapter dies before ever
  // answering. `error` (e.g. spawn ENOENT: the binary doesn't exist) feeds
  // the same channel: Node's ChildProcess throws if an 'error' event has no
  // listener at all, so without this a missing binary crashes the whole
  // process instead of failing this session cleanly.
  const childClosed = new Promise<never>((_resolve, reject) => {
    child.once("close", (code, signal) => {
      reject(new Error(`claude-code-acp exited (code=${code} signal=${signal}): ${stderrTail.trim()}`));
    });
    child.once("error", (err) => {
      reject(new Error(`failed to launch claude-code-acp: ${err.message}`));
    });
  });
  childClosed.catch(() => {}); // don't warn if nothing ever awaits this

  // Tracks the session's mode state; kept live so `session.modes` is always
  // current, whether it changed via our own setMode() or the agent's own
  // current_mode_update notification.
  let modes: SessionModeState | null = null;

  const client: Client = {
    async sessionUpdate(params) {
      const u = toUpdate(opts.cwd, params);
      if (u.kind === "current_mode_update" && u.modeId != null && modes != null) {
        modes = { ...modes, currentModeId: u.modeId };
      }
      opts.onUpdate(u);
    },
    async requestPermission(params: RequestPermissionRequest): Promise<RequestPermissionResponse> {
      const optionId = await opts.onPermission({
        title: permissionTitle(params.toolCall),
        options: params.options,
        toolCall: params.toolCall,
      });
      return {
        outcome: optionId === null ? { outcome: "cancelled" } : { outcome: "selected", optionId },
      };
    },
  };

  const conn = new ClientSideConnection(
    () => client,
    ndJsonStream(Writable.toWeb(child.stdin), Readable.toWeb(child.stdout)),
  );

  // Signals `signal` to the whole process group `child` leads (see spawn's
  // detached above), not just `child` itself — that is what actually reaches
  // a grandchild like the real agent process npx execs into. ESRCH (nothing
  // left to signal) is success, not an error.
  function killTree(signal: NodeJS.Signals): void {
    try {
      if (process.platform !== "win32" && child.pid != null) {
        process.kill(-child.pid, signal);
      } else {
        child.kill(signal);
      }
    } catch (err) {
      if ((err as NodeJS.ErrnoException).code !== "ESRCH") throw err;
    }
  }

  // conn.closed resolves when the ndjson *stream* ends — our end of the
  // pipe — not when the process group exits. argv is commonly a wrapper
  // (npx) execing a shell execing the real agent; the wrapper's own hold on
  // that pipe can end well before the tree it spawned actually dies (it can
  // even close its stdout without dying at all). So conn.closed is never
  // trusted as proof of death here — the process group itself, polled
  // directly, is. process.kill(-pid, 0) throws ESRCH once nothing is left in
  // the group to signal; any other outcome (success, or a different errno
  // like EPERM) means something is still there. Win32 has no process-group
  // signal, so there we fall back to the immediate child's own "exit" —
  // the same platform limitation killTree already documents.
  let childExited = false;
  child.once("exit", () => {
    childExited = true;
  });

  function groupAlive(): boolean {
    if (process.platform === "win32") return !childExited;
    if (child.pid == null) return false;
    try {
      process.kill(-child.pid, 0);
      return true;
    } catch (err) {
      return (err as NodeJS.ErrnoException).code !== "ESRCH";
    }
  }

  // Polls groupAlive() until it's false or deadlineMs elapses. 50ms is short
  // enough that a process which dies promptly (the common case) doesn't add
  // noticeable latency to close().
  async function waitForGroupGone(deadlineMs: number): Promise<boolean> {
    const deadline = Date.now() + deadlineMs;
    while (groupAlive()) {
      if (Date.now() >= deadline) return false;
      await delay(50);
    }
    return true;
  }

  let closed = false;
  // Never hangs and never rejects: idempotent via `closed`, and bounded by
  // the grace period plus the post-SIGKILL confirmation wait below,
  // regardless of what the child (or a stubborn descendant that traps
  // SIGTERM) does.
  async function close(): Promise<void> {
    if (closed) return;
    closed = true;

    killTree("SIGTERM");

    if (!(await waitForGroupGone(1000))) {
      killTree("SIGKILL");
      // SIGKILL cannot be ignored, so the group is coming down; confirm it's
      // actually gone (bounded) before resolving, rather than just assuming
      // the signal worked.
      await waitForGroupGone(2000);
    }
  }

  let sessionId: string;
  try {
    // Raced the same way prompt() is raced against childClosed, and for the
    // same underlying reason: the SDK never rejects a pending request on its
    // own. Without this, an adapter that exits (or never spawns, or never
    // answers) during startup leaves initialize()/newSession() pending
    // forever — a blank terminal and a leaked worktree/branch, with nothing
    // ever printed and the caller's own catch never reached.
    const initFlow = (async () => {
      await conn.initialize({
        protocolVersion: PROTOCOL_VERSION,
        clientCapabilities: { fs: { readTextFile: false, writeTextFile: false } },
      });
      return conn.newSession({ cwd: opts.cwd, mcpServers: [] });
    })();

    const result = await Promise.race([
      initFlow,
      childClosed,
      startupTimeout(opts.startupTimeoutMs ?? DEFAULT_STARTUP_TIMEOUT_MS),
    ]);
    sessionId = result.sessionId;
    modes = result.modes ?? null;
  } catch (err) {
    // Startup failed one way or another: nothing here is usable, so don't
    // leave the child (or, if it's still alive — the hung-adapter/timeout
    // case — its whole process group) running unattended.
    await close().catch(() => {});
    throw err;
  }

  return {
    sessionId,
    get modes(): SessionModeState | null {
      return modes;
    },
    async setMode(modeId: string): Promise<void> {
      await conn.setSessionMode({ sessionId, modeId });
      // The response carries no state; a successful call means the agent is
      // now in this mode, so update optimistically rather than wait for a
      // current_mode_update that may not come.
      if (modes != null) modes = { ...modes, currentModeId: modeId };
    },
    async prompt(text: string): Promise<string> {
      const { stopReason } = await Promise.race([
        conn.prompt({ sessionId, prompt: [{ type: "text", text }] }),
        childClosed,
      ]);
      return stopReason;
    },
    async cancel(): Promise<void> {
      await conn.cancel({ sessionId });
    },
    close,
  };
}
